uc-crux-llvm/src/UCCrux/LLVM/Context/Function.hs

```diff
@@ -0,0 +1,275 @@
+      unimplemented "makeFunctionContext" VarArgsFunction
+    let trans = modCtx ^. moduleTranslation
+    funDecl <-
+      withTypeContext modCtx $
+        case LLVMTrans.liftDeclare (LLVMTrans.declareFromDefine def) of
+          Left err -> Left (BadLift (Text.pack err))
+          Right d -> Right d
+    argMemTypes <-
+      case maybeMapToContext
+        (Ctx.size argTypes)
+        (Map.fromList (zip [0 ..] (fdArgTypes funDecl))) of
+        Right types -> Right types
+        Left idx -> Left (BadLiftArgs (length (fdArgTypes funDecl)) (Ctx.sizeInt (Ctx.size argTypes)) idx)
+    ( SomeAssign
+        (argFullTypes :: Ctx.Assignment (FullTypeRepr m) fullTypes)
+        (mts :: ModuleTypes m)
+        (Refl :: MapToCrucibleType arch fullTypes :~: argTypes)
+      ) <-
+      withTypeContext modCtx $
+        case assignmentToFullType trans argTypes argMemTypes of
+          Right fullTypes -> Right fullTypes
+          Left ident -> Left (FullTypeTranslation ident)
+    argMemTypes' <-
+      case maybeMapToContext
+        (Ctx.size argFullTypes)
+        (Map.fromList (zip [0 ..] (fdArgTypes funDecl))) of
+        Right types -> Right types
+        Left idx -> Left (BadLiftArgs (length (fdArgTypes funDecl)) (Ctx.sizeInt (Ctx.size argFullTypes)) idx)
+    argStorageTypes <-
+      withPtrWidthOf trans $
+        forFC
+          argMemTypes'
+          ( \(Const memType) ->
+              case toStorableType memType of
+                Just storeTy -> Right (Const storeTy)
+                Nothing -> Left (BadMemType memType)
+          )
+    pure $
+      SomeFunctionContext
+        ( FunctionContext
+            { _functionName = name,
+              _moduleTypes = mts,
+              _argumentFullTypes = argFullTypes,
+              _argumentCrucibleTypes = argTypes,
+              _argumentMemTypes = argMemTypes',
+              _argumentStorageTypes = argStorageTypes,
+              _argumentNames =
+                fmapFC
+                  (Const . getFirst . getConst)
+                  ( mapToContext
+                      (Ctx.size argFullTypes)
+                      (fmap (First . Just) (debugInfoArgNames llvmMod def))
+                  )
+            }
+        )
+        Refl
+
+mapToContext ::
+  Monoid a =>
+  Ctx.Size items ->
+  Map Int a ->
+  Ctx.Assignment (Const a) items
+mapToContext size mp =
+  Ctx.generate
+    size
+    (\index -> Const (Map.findWithDefault mempty (Ctx.indexVal index) mp))
+
+maybeMapToContext ::
+  Ctx.Size items ->
+  Map Int a ->
+  Either Int (Ctx.Assignment (Const a) items)
+maybeMapToContext size mp =
+  traverseFC (fmap Const . getConst) $
+    Ctx.generate
+      size
+      ( \index ->
+          Const
+            ( case Map.lookup (Ctx.indexVal index) mp of
+                Just val -> Right val
+                Nothing -> Left (Ctx.indexVal index)
+            )
+      )
+
+-- Stolen shamelessly from saw-script
+debugInfoArgNames :: L.Module -> L.Define -> Map Int Text
```

Comment:
Should we lift this somewhere else?

Reply:
I'm open to suggestions! Happy to move it, I just don't know where it should go

Reply:
Good question.  Maybe it should lift all the way into `llvm-pretty`, or perhaps we should build a utility package on top of it for higher-level parsing of debug info.  Regardless, I suppose it's a bit incidental to this PR, but we should probably open a ticket about this code duplication and see if it makes sense to do something better.

Reply:
https://github.com/GaloisInc/crucible/issues/696